shopfloor_base/utils.py

```diff
@@ -65,3 +71,66 @@ def _get_running_env():
 
 
 RUNNING_ENV = _get_running_env()
+
+
+def register_new_services(env, *component_classes, apps=None):
+    """Enforce registration of new services straight from their classes.
+
+    At some stage (eg: a post install hook) the component registry
+    might not be fully loaded and the components might be not fully "ready".
+    For this reason, not all the inherited properties
+    are granted to be available (eg: _collection)
+    and they won't be available for `components_registry.lookup` call.
+
+    This is why, to ensure that the new services endpoints
+    are registered in any case we prepare their classes "manually" here
+    and pass them over to sf.app._get_services.
+
+    In this way, if their endpoint_route rows are not there
+    they will be generated on the fly,
+
+    Please, note that this func is very low level
+    and must be used only in particular situations (such as install hooks).
+    """
+    if odoo_config["test_enable"]:
+        # When installing modules for tests, there's nothing to do.
+        # Moreover, it can lead to weird service behavior
+        # because the classes will be decorated
+        # but when they are not fully ready.
+        return
+    components = load_components_without_registry(env, *component_classes)
+    if not components:
+        return
+    apps = apps or env["shopfloor.app"].search([])
+    apps.with_context(sf_service_components=components)._register_controllers()
+
+
+def load_components_without_registry(env, *component_classes):
+    """Prepare components' instances w/o real registry lookup."""
+    collection = _PseudoCollection("shopfloor.app", env)
+    rest_service = env["rest.service.registration"]
+    if not _component_databases.get(env.cr.dbname):
+        # no registry ready yet (eg: install via cmd line, tests, etc)
+        _logger.info("component registry not ready yet")
+        return []
+    work = WorkContext(model_name=rest_service._name, collection=collection)
+    return [comp(work) for comp in component_classes]
+
+
+def purge_endpoints(env, service_usage):
+    """Remove stale services' endpoints routes.
+
+    When scenario are removed (eg: module uninstalled)
+    their routes must be cleaned up.
+    You can use this func to easily clean them app for all apps.
+
+    Motivation: registered routes for a given app are taken
+    from the endpoint_route table via `route_group`.
+    As the table is populated dynamically with registered components
+    when a component is removed there's no direct update to the table.
+    Hence, is the responsibility of the dev to take care of the cleanup.
+    """
+    apps = env["shopfloor.app"].search([])
+    for app in apps:
+        route = app.api_url_for_service(service_usage)
+        env.cr.execute("DELETE FROM endpoint_route WHERE route like %s", (f"{route}%",))
```

Comment:
```suggestion
        env.cr.execute("DELETE FROM endpoint_route WHERE route like %s", (f"{route}/%",))
```
I think the trailing slash is not included, this would be safer.

Also this function will not work for Odoo modules that extend an existing shopfloor scenario with new route.

Reply:
What do you mean? A new route will still be under the same base route of the scenario.

Reply:
What if one day there is two services, the 2nd one starting with the name of the first one ('reception', `reception_something' ?

Reply:
For the 2nd point I was thinking that this implementation can not be used for some module like in https://github.com/OCA/wms/pull/647 for the module `shopfloor_reception_packaging_dimension` but that is not blocking I guess just a warning

Reply:
Good catches, you are right. Please check my last commit: 
1. use "/" for main routes
2. accept "endpoint" if you want to affect only specific endpoints

Untested: could you give it a try for #647?

Reply:
Tested and it works, with a little gotcha, when importing a service that extends an existing component `_usage`, `_name` are not set. More info in #647 when I push the fix